src/QirRuntime/lib/QIR/intrinsics.cpp

```diff
@@ -15,38 +15,79 @@
+    for (long i = 0; i < count; i++)
+    {
+        pauliIds.push_back(static_cast<PauliId>(*paulis->GetItemPointer(i)));
+    }
+    return pauliIds;
+}
+
 extern "C"
 {
-    double quantum__qis__intAsDouble(long value)
+    void quantum__qis__exp__body(QirArray* paulis, double angle, QirArray* qubits)
+    {
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->Exp(
+            paulis->count, reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer),
+            angle);
+    }
+
+    void quantum__qis__exp__adj(QirArray* paulis, double angle, QirArray* qubits)
+    {
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->Exp(
+            paulis->count, reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer),
+            -angle);
+    }
+
+    void quantum__qis__exp__ctl(QirArray* ctls, QirArray* paulis, double angle, QirArray* qubits)
     {
-        assert(value == static_cast<int32_t>(value));
-        return static_cast<double>(value);
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->ControlledExp(
+            ctls->count, reinterpret_cast<Qubit*>(ctls->buffer), paulis->count,
+            reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer), angle);
     }
 
-    void quantum__qis__cnot(Qubit control, Qubit qubit)
+    void quantum__qis__exp__ctladj(QirArray* ctls, QirArray* paulis, double angle, QirArray* qubits)
     {
-        Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->ControlledX(1, &control, qubit);
+        assert(paulis->count == qubits->count);
+
+        std::vector<PauliId> pauliIds = ExtractPauliIds(paulis);
+        return Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->ControlledExp(
+            ctls->count, reinterpret_cast<Qubit*>(ctls->buffer), paulis->count,
+            reinterpret_cast<PauliId*>(pauliIds.data()), reinterpret_cast<Qubit*>(qubits->buffer), angle);
     }
 
-    void quantum__qis__h(Qubit qubit)
+    void quantum__qis__h__body(Qubit qubit)
     {
         Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->H(qubit);
     }
 
-    Result quantum__qis__measure(QirArray* paulis, QirArray* qubits)
+    void quantum__qis__h__ctl(QirArray* ctls, Qubit qubit)
+    {
+        Microsoft::Quantum::g_context->simulator->AsQuantumGateSet()->ControlledH(
+            ctls->count, reinterpret_cast<Qubit*>(ctls->buffer), qubit);
+    }
+
+    Result quantum__qis__measure__body(QirArray* paulis, QirArray* qubits)
     {
         const long count = qubits->count;

```

Comment:
Very minor: this is the one function where you save the qubit count into a local variable to avoid calling it twice (once in the assert and once in the simulator invocation). Is it worth removing this to keep it looking like the other methods?

Reply:
I have a mild preference for how it is now as it keeps the line of arguments from wrapping one more time...